nutils/evaluable.py

```diff
@@ -1656,139 +1657,154 @@ def __init__(self, funcs: types.frozenmultiset):
-            return d1 * d2
-        unaligned2, where2 = unalign(func2)
-        if {axis1, axis2}.isdisjoint(where2):
-            d1 = determinant(func1, (axis1, axis2))
-            d2 = align(unaligned2**self.shape[axis1], [i-(i > axis1)-(i > axis2) for i in where2 if i not in (axis1, axis2)], d1.shape)
-            return d1 * d2
+        factors = tuple(self._factors)
+        if all(_equals_scalar_constant(self.shape[axis], 1) for axis in (axis1, axis2)):
+            return multiply(*[determinant(f, (axis1, axis2)) for f in factors])
+        for i, fi in enumerate(factors):
+            unaligned, where = unalign(fi)
+            if axis1 not in where and axis2 not in where:
+                det = determinant(multiply(*factors[:i], *factors[i+1:]), (axis1, axis2))
+                scale = align(unaligned**self.shape[axis1], [i-(i > axis1)-(i > axis2) for i in where if i not in (axis1, axis2)], det.shape)
+                return det * scale
 
     def _product(self):
-        func1, func2 = self.funcs
-        return multiply(Product(func1), Product(func2))
-
-    def _multiply(self, other):
-        func1, func2 = self.funcs
-        func1_other = func1._multiply(other)
-        if func1_other is not None:
-            return multiply(func1_other, func2)
-        func2_other = func2._multiply(other)
-        if func2_other is not None:
-            return multiply(func1, func2_other)
+        return multiply(*[Product(f) for f in self._factors])
 
     def _derivative(self, var, seen):
         func1, func2 = self.funcs
         return einsum('A,AB->AB', func1, derivative(func2, var, seen)) \
             + einsum('A,AB->AB', func2, derivative(func1, var, seen))
 
     def _takediag(self, axis1, axis2):
-        func1, func2 = self.funcs
-        return multiply(_takediag(func1, axis1, axis2), _takediag(func2, axis1, axis2))
+        return multiply(*[_takediag(f, axis1, axis2) for f in self._factors])
 
     def _take(self, index, axis):
-        func1, func2 = self.funcs
-        return multiply(_take(func1, index, axis), _take(func2, index, axis))
+        return multiply(*[_take(f, index, axis) for f in self._factors])
 
     def _sign(self):
-        func1, func2 = self.funcs
-        return multiply(Sign(func1), Sign(func2))
+        return multiply(*[Sign(f) for f in self._factors])
 
     def _unravel(self, axis, shape):
-        func1, func2 = self.funcs
-        return multiply(unravel(func1, axis, shape), unravel(func2, axis, shape))
+        return multiply(*[unravel(f, axis, shape) for f in self._factors])
 
     def _inverse(self, axis1, axis2):
-        func1, func2 = self.funcs
-        if set(unalign(func1)[1]).isdisjoint((axis1, axis2)):
-            return divide(inverse(func2, (axis1, axis2)), func1)
-        if set(unalign(func2)[1]).isdisjoint((axis1, axis2)):
-            return divide(inverse(func1, (axis1, axis2)), func2)
+        factors = tuple(self._factors)
```

Comment:
Just a thought: With [more_itertools.partition](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.partition) we can split factors in those that have `axis1` or `axis2` and those that have not:

```python
factors_without, factors_with = more_itertools.partition(
    lambda factor: set(unalign(factor)[1]).isdisjoint((axis1, axis2)),
    factors)
```

Then proceed with 

```python
if factors_with and factors_without:
    divide(inverse(multiply(*factors_with)), multiply(*factors_without))
```

Reply:
But do we want to pull in a dependency for that? With the code as it is, all "without" factors will eventually come out, it will just take a few iterations. Given that there is often at most one (scalar multiplications are merged via InsertAxis._multiply) I think that's fine.